source/Calamari.Shared/Integration/Processes/SilentProcessRunner.cs

```diff
@@ -149,11 +152,27 @@ static SilentProcessRunner()
                         process.BeginOutputReadLine();
                         process.BeginErrorReadLine();
 
-                        process.WaitForExit();
-                        outputWaitHandle.WaitOne();
-                        errorWaitHandle.WaitOne();
+                        // Some processes can have race conditions. Between the call to Start and WaitForExit part of the process may have exited already.
+                        // This can happen when calling CommitChanges in dotnet's ServerManager as shown here: https://stackoverflow.com/questions/7446632/servermanager-commitchanges-makes-changes-with-a-slight-delay
+                        // Commit changes is called by appcmd from IIS, which is called by Set-ItemProperty from the WebAdministration module in Powershell
+                        // https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.waitforexit?view=netframework-4.8#System_Diagnostics_Process_WaitForExit_System_Int32_
+                        // Add a timeout passed down from some default configuration or context up the call stack, -1 will never timeout.
+                        var timedOut = !process.WaitForExit(timeoutMilliseconds);
 
-                        return new SilentProcessRunnerResult(process.ExitCode, errorData.ToString());
+                        if (!timedOut)
+                        {
+                            // Only wait when the process has not timed out, otherwise we will be stuck here as well.
+                            outputWaitHandle.WaitOne();
```

Comment:
I think we still want to do this, as it'll finish the reading from the console. Would just adding the following work?
```
if(!processExited)
{
    Log.Error($"Process with ID {process.Id} exceeded the max allowed runtime of {timeoutMilliseconds} milliseconds and will be killed.");
    process.Kill();
    process.WaitForExit();
}
```

Reply:
Works much better and cleaner to read. Much obliged :)